pyccel/ast/datatypes.py

```diff
@@ -12,115 +12,201 @@
+    """
+    Class representing a complex datatype.
+
+    Class representing a complex datatype.
+    """
+    __slots__ = ()
+    _name = 'complex'
+
+class PyccelCharacterType(PrimitiveType):
+    """
+    Class representing a character datatype.
+
+    Class representing a character datatype.
+    """
+    __slots__ = ()
+    _name = 'character'
 
 #==============================================================================
 
-class DataType:
+class PyccelType:
     """
-    Base class representing native datatypes.
+    Base class representing the type of an object.
+
+    Base class representing the type of an object from which all
+    types must inherit. A type must contain enough information to
+    describe the declaration type in a low-level language.
+
+    Types contain an addition operator. The operator indicates the type that
+    is expected when calling an arithmetic operator on objects of these types.
 
-    The base class from which all data types must inherit.
+    Where applicable, types also contain an and operator. The operator indicates the type that
+    is expected when calling a bitwise comparison operator on objects of these types.
     """
     __slots__ = ()
-    _name = '__UNDEFINED__'
 
     def __init__(self): #pylint: disable=useless-parent-delegation
         # This __init__ function is required so the ArgumentSingleton can
         # always detect a signature
         super().__init__()
 
+    def __str__(self):
+        return self._name #pylint: disable=no-member
+
+    def switch_basic_type(self, new_type):
+        """
+        Change the basic type to the new type.
+
+        Change the basic type to the new type. In the case of a FixedSizeType the
+        switch will replace the type completely, directly returning the new type.
+        In the case of a homogeneous container type, a new container type will be
+        returned whose underlying elements are of the new type. This method is not
+        implemented for inhomogeneous containers.
+
+        Parameters
+        ----------
+        new_type : PyccelType
+            The new basic type.
+
+        Returns
+        -------
+        PyccelType
+            The new type.
+        """
+        raise NotImplementedError(f"switch_basic_type not implemented for {type(self)}")
+
+#==============================================================================
+
+class FixedSizeType(PyccelType, metaclass=Singleton):
+    """
+    Base class representing a built-in scalar datatype.
+
+    The base class representing a built-in scalar datatype which can be
+    represented in memory. E.g. int32, int64.
+    """
+    __slots__ = ()
+
     @property
-    def name(self):
+    def datatype(self):
         """
-        Get the name of the datatype.
+        The datatype of the object.
 
-        Get the name of the datatype.
+        The datatype of the object.
         """
-        return self._name
+        return self
```

Comment:
What is the purpose of this property?

Reply:
Similarly to the discussion of `switch_basic_type` below, the purpose of this property is to facilitate access to the datatype without requiring if statements in lots of places across the code. For a `FixedSizeType` this property seems useless as it just returns self however if we have a `list[tuple[int,..]]` we can get the underlying datatype (int) by doing `class_type.datatype` this will go through the following call graph:

`HomogeneousListType(HomogeneousTupleType(PythonNativeInt())).datatype` -> `HomogeneousTupleType(PythonNativeInt()).datatype` -> `PythonNativeInt().datatype`

using this function:
https://github.com/pyccel/pyccel/blob/889193ffc64ca97ab44cb5c7264f0bd682b18c9d/pyccel/ast/datatypes.py#L470-L477

Reply:
I see. So the scalar class returns `self` to stop the recursive calls and provide a usable result.

Reply:
The relevant part of the docs is here:
https://github.com/pyccel/pyccel/blob/889193ffc64ca97ab44cb5c7264f0bd682b18c9d/developer_docs/type_inference.md#L75